### ISSUE TITLE
Review how symlinks are handled

### ISSUE THREAD

**Issue by aeqz** [serokell/xrefcheck#244]
# Clarification and motivation

After #197, we have decided to not use `System.Directory.CanonicalizePaths` anymore, and to use a weaker custom version instead. With that change, we lost the automatic symlink resolution that `System.Directory.CanonicalizePaths` does, so we should think about how to handle them if necessary.

# Acceptance criteria

Do some tests involving symlinks and reason about if we need to handle them in some way and, in that case, how we could do it.

**Comment by aeqz** [comment]
I have tried the example from #242 and the result is the same after #197:

```
=== Invalid references found ===                                                                                                                    

  ➥  In file s.md
     bad reference (relative) at src:1:1-9:
       - text: "Link"
       - link: a
       - anchor: -

     File does not exist:
       a

Invalid references dumped, 1 in total.
```

**Comment by Martoon-00** [comment]
Well, generally I think it's clear what we want: see how GitHub/GitLab do this, and handle it by analogy.

### From the point of view "file with references"

If attempt to open a symlink to markdown document causes them to open the target document, then there are two cases:
* If the symlink and the target reside in the same repository - then it's an interesting question. Probably, if target is checked and not ignored, we can skip checking the symlink.
* If they reside in different directories, if the target contains some references, then we can expect links to be 100% wrong (assuming in the target file they are valid), and we can display the respective message saying that symlink cannot contain valid references.

If attempt to open a symlink to markdown document causes the site to open symlink as symlink (e.g. GitLab seems to do so, but let's extra check how it behaves in different versions of GitLab), then we shouldn't check that link.

Maybe we actually could implement a separate type of scanner for symlinks - I'm not sure how good will this work. If you try it and fail, please write down which problems you faced and leave a branch that demostrates your effort.

### From the point of view "file exposing anchors"

Here we also should check how GitHub/GitLab behave.

**Comment by aeqz** [comment]
I have tried how symlinks are rendered on GitHub, and the behaviour is the same as on GitLab: they just render a label with the path of the file they point to. Also, at first sight, I have not found any information showing that this was different in a previous GitLab version.

I think that it would be interesting to check at least that symlinks point to an existing file, but I'm not sure if it would be problematic in case that someone has intentionally any symlink pointing outside the repository. Some configuration should be required then, either globally or with some kind of ignore list.

**Comment by Martoon-00** [comment]
> I think that it would be interesting to check at least that symlinks point to an existing file

That's all an interesting idea.

I agree that this may be worth handling. In this case, I think, it would make sense to try to add a completely separate scanner for this, have a dedicated config for it in `scanners` config section, e.t.c.

Also, if GitHub and GitLab do not bother themselves with symlinks, then the question is - what is the correct symlink? We seem to have two answers to this question - one from Linux, another from Windows :smile:   Perhaps just assuming that symlink format matches the developers' machines and machines the CI machine is fine.

I guess, let's create a ticket for this. This is not necessary for 0.3.0 because it's a new feature, not a bug fix, but this seems like something worthy to have.

**Comment by aeqz** [comment]
I have removed this issue from 0.3.0 and added #242, which is the bug related with symlinks that is fixed by #261, so we can deal with the symlink scanner in this same issue. Does it work?

**Comment by Martoon-00** [comment]
Sounds good, let's go so :+1:
